### PR TITLE
`r/azurerm_xyz_policy_assignment` add support for `non_compliance_message`

### LINKED ISSUE #10401: Support for nonComplianceMessages in azurerm_policy_assignment
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

API version 2020-09-01 for policy assignment supports specifying custom message (properties.nonComplianceMessages) so users can get clear understanding of why resource is non compliant. Would be great to add this to azurerm_policy_assignment.

### New or Affected Resource(s)

azurerm_policy_assignment

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

https://docs.microsoft.com/en-us/rest/api/resources/policyassignments/create

* #0000

### PR DESCRIPTION
**Pull request by matt-FFFFFF**
### This PR:

* Adds support for Non-Compliance Messages to:
  * `r/azurerm_subscription_policy_assignment`
  * `r/azurerm_resource_group_policy_assignment`
  * `r/azurerm_management_group_policy_assignment`
  * `r/azurerm_resource_policy_assignment`
* Update policy API version to `2021-06-01-preview`
* fixes #10401 

### Acceptance tests

<details><summary>Acceptance Tests for Policies</summary><p>

---
```text
❯ TF_ACC=1 /usr/local/go/bin/go test -v ./internal/services/policy -run=TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
=== PAUSE TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
=== CONT  TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
--- PASS: TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage (843.31s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/policy	844.474s

❯ TF_ACC=1 /usr/local/go/bin/go test -v ./internal/services/policy -run=TestAccResourcePolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccResourcePolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
=== PAUSE TestAccResourcePolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
=== CONT  TestAccResourcePolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
--- PASS: TestAccResourcePolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage (860.95s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/policy	862.068s

❯ TF_ACC=1 /usr/local/go/bin/go test -v ./internal/services/policy -run=TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
=== PAUSE TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
=== CONT  TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
--- PASS: TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage (848.64s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/policy	850.804s

❯ TF_ACC=1 /usr/local/go/bin/go test -v ./internal/services/policy -run=TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
=== PAUSE TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
=== CONT  TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage
--- PASS: TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicyNonComplianceMessage (689.18s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/policy	690.368s
```
---
</p></details>

<details><summary>Acceptance Tests for Policy Sets</summary><p>

---
```text
❯ TF_ACC=1 /usr/local/go/bin/go test -v ./internal/services/policy -run=TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
=== PAUSE TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
=== CONT  TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
--- PASS: TestAccSubscriptionPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage (819.12s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/policy	820.257s

❯ TF_ACC=1 /usr/local/go/bin/go test -v ./internal/services/policy -run=TestAccResourcePolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccResourcePolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
=== PAUSE TestAccResourcePolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
=== CONT  TestAccResourcePolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
--- PASS: TestAccResourcePolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage (1052.03s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/policy	1056.193s

❯ TF_ACC=1 /usr/local/go/bin/go test -v ./internal/services/policy -run=TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
=== PAUSE TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
=== CONT  TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
--- PASS: TestAccResourceGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage (949.83s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/policy	952.996s

❯ TF_ACC=1 /usr/local/go/bin/go test -v ./internal/services/policy -run=TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
=== PAUSE TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
=== CONT  TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage
--- PASS: TestAccManagementGroupPolicyAssignment_basicWithBuiltInPolicySetNonComplianceMessage (943.61s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/policy	945.769s
```
---
</p></details>

The following has been done:

* [x] Documentation updated
* [x] Acceptance tests updated
* [x] Linting passed

### CODE REVIEW

File: internal/services/policy/assignment_base_resource.go [hunk 1]
```diff
@@ -326,6 +336,26 @@ func (br assignmentBaseResource) arguments(fields map[string]*pluginsdk.Schema)
 			},
 		},
 
+		"non_compliance_message": {
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"message": {
```
Comment (katbyte) [inline internal/services/policy/assignment_base_resource.go:344]: as this is part of the block already should we make this
```suggestion
					"content": {
```

File: internal/services/policy/assignment_base_resource.go [hunk 2]
```diff
@@ -326,6 +336,26 @@ func (br assignmentBaseResource) arguments(fields map[string]*pluginsdk.Schema)
...
+						Type:         pluginsdk.TypeString,
+						Required:     true,
+						ValidateFunc: validation.StringIsNotEmpty,
+					},
+
+					"policy_definition_reference_id": {
+						Type:         pluginsdk.TypeString,
+						Optional:     true,
+						ValidateFunc: validation.StringIsNotEmpty,
```
Comment (katbyte) [inline internal/services/policy/assignment_base_resource.go:353]: and could we validate this with the policy reference id validator?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @matt-FFFFFF - couple comments about the schema whilst passing through

File: internal/services/policy/assignment_base_resource.go (hunk 1 above)
Comment (matt-FFFFFF) [reply internal/services/policy/assignment_base_resource.go:344]: Hi - thanks for looking at this.

I went for `message` as this matched the ARM REST schema:

https://docs.microsoft.com/en-us/azure/governance/policy/concepts/assignment-structure#non-compliance-messages

Happy to change it if you want but felt this was analogous to the source. Please let me know!

File: internal/services/policy/assignment_base_resource.go (hunk 2 above)
Comment (matt-FFFFFF) [reply internal/services/policy/assignment_base_resource.go:353]: Hi,

This isn't a resource ID, this is a string value to identity each policy definition within a set. For an example see here:

https://docs.microsoft.com/en-us/azure/governance/policy/concepts/assignment-structure#non-compliance-messages

And here:

https://docs.microsoft.com/en-us/azure/templates/microsoft.authorization/policysetdefinitions?tabs=bicep#policydefinitionreference

Where it's defined as:

policyDefinitionReferenceId - A unique id (within the policy set definition) for this policy definition reference. (string)

Comment (matt-FFFFFF) [reply internal/services/policy/assignment_base_resource.go:353]: Resolving for now as `string` is correct here

File: internal/services/policy/assignment_base_resource.go (hunk 1 above)
Comment (katbyte) [reply internal/services/policy/assignment_base_resource.go:344]: we try to be more clear the accurate to the API as the resulting UX can be less than optimal having message.message is less clear and ambiguous then message.content 🙂

Comment (matt-FFFFFF) [reply internal/services/policy/assignment_base_resource.go:344]: Sure thing, will update now 👍

Comment (matt-FFFFFF) [reply internal/services/policy/assignment_base_resource.go]: Resolved

**Comment by matt-FFFFFF** [comment]
> Thanks @matt-FFFFFF - couple comments about the schema whilst passing through 

Hi @katbyte,

I've implemented the changes now and the checks pass. Are you able to take another look please?

😄

**Review by mbfrahry — APPROVED** [review]
LGTM! The tests all pass and I really appreciate testing across all the policies we are touching here. Thanks @matt-FFFFFF!

**Pull request merged**